usb/esp_tinyusb/Kconfig

```diff
@@ -160,4 +160,66 @@ menu "TinyUSB Stack"
             help
                 Setting value greater than 0 will enable TinyUSB HID feature.
     endmenu # "HID Device Class (HID)"
+
+    menu "Mass Storage Device Class (MSC)"
+        choice STORAGE_MEDIA
```

Comment:
Can this be decided at link time and run time, instead of configuration time? Some of the projects (Arduino, Micropython) use ESP-IDF and other components in a pre-compiled form, therefore fixing this choice here will mean that they would need to build two versions of a precompiled library.

Could we introduce two functions,

```
esp_err_t tinyusb_msc_storage_init_sdcard(const tinyusb_msc_storage_config_t *config, more args);
esp_err_t tinyusb_msc_storage_init_spiflash(const tinyusb_msc_storage_config_t *config more args);
```

instead?

Reply:
I have now taken care of it.
Now, the APIs that are defined has one more argument - `const tinyusb_msc_storage_type_t storage_type`
This parameter will tell this driver to choose and operate on eaither spiflash or sd-card. Later the API usage can be extended for more media types without changing the API signature.

Reply:
I have now taken care of it.
Now, the APIs that are defined has one more argument - `const tinyusb_msc_storage_type_t storage_type`
This parameter will tell this driver to choose and operate on eaither spiflash or sd-card. Later the API usage can be extended for more media types without changing the API signature.

Reply:
If you do this at run time based on storage_type argument, this means that both the SDMMC and Flash driver will be linked into the application, even if only one of the two is ever used.

Having two separate functions avoids the problem — the linker will be able to eliminate the code that is not called.

Reply:
Currently, we have 7 common APIs for spi-flash and sdcard.
If we separate out APIs for spiflash and sdcard, it will be 14 APIs. In future, if more media type is supported, there will be an increase of number of APIs. The code size will increase with more files.
I request you check the current implementation once more.

Reply:
In IDF, this dependency issue is usually solved using a factory method pattern, something along these lines:

```c++
// header file:


// forward declaration of the opaque handle
typedef struct mylibrary_handle_s *mylibrary_handle_t;

// factory method for implementation 1:
esp_err_t mylibrary_init_impl_aaa(const mylibrary_config_aaa_t* config, mylibrary_handle_t* out_handle);

// factory method for implementation 2:
esp_err_t mylibrary_init_impl_bbb(const mylibrary_config_bbb_t* config, mylibrary_handle_t* out_handle);

// common method for every implementation
esp_err_t mylibrary_do_stuff(mylibrary_handle_t handle, int some_arg);
```
```c++
// source file:

struct mylibrary_handle_s {
    // implementation function pointer
    esp_err_t (*do_stuff_impl)(int);
}

esp_err_t mylibrary_do_stuff(mylibrary_handle_t handle, int some_arg) {
    // indirect call
    return (handle->do_stuff_impl)(some_arg);
}

static esp_err_t mylibrary_aaa_do_stuff(mylibrary_handle_t handle, int some_arg)
{
    // concrete implementation for aaa
}

static esp_err_t mylibrary_bbb_do_stuff(mylibrary_handle_t handle, int some_arg)
{
    // concrete implementation for bbb
}

esp_err_t mylibrary_init_impl_aaa(const mylibrary_config_aaa_t* config, mylibrary_handle_t* out_handle)
{
    mylibrary_handle_t handle = (allocate struct mylibrary_handle_s, error check...)
    handle->do_stuff_impl = &mylibrary_aaa_do_stuff;
    *out_handle = handle;
    return ESP_OK;
}

esp_err_t mylibrary_init_impl_bbb(const mylibrary_config_bbb_t* config, mylibrary_handle_t* out_handle)
{
    mylibrary_handle_t handle = (allocate struct mylibrary_handle_s, error check...)
    handle->do_stuff_impl = &mylibrary_bbb_do_stuff;
    *out_handle = handle;
    return ESP_OK;
}
```

This way, only those implementations of functions which are actually needed are linked into the final binary. You can also easily extend the implementation to new subclasses by adding more factory methods and implementations.